### DISCUSSION TITLE
VSGB & VSGT format specification

### DISCUSSION THREAD

**Discussion by shinyclaw** [vsg-dev/VulkanSceneGraph#1157] [General]
Hi, I want to write a custom fbx to VSG native converter based on https://ufbx.github.io/. Where I can find the specification of **vsgb** and **vsgt** files?

**Comment by robertosfield** [comment]
The native formats are meant for serializing in/out VulkanSceneGraph objects without being tied to constraints of standardized interchange format.

Each vsg::Object subclass defines how it read/write it's members, and each object's read/write methods are responsible for versioning.  This means the format isn't a public property rather is external specified like you'd find for a standardized format meant for data interchange like glTF.

If you need to read/write to native VSG formats I'd recommend linking to the VSG, writing a loader that creates a VSG scene graphs/object hierarchy then using vsg::write(..) to create the final .vsgb or .vsvgt file.  This way you can focus on setting up the scene graph in an appropriate way for rendering and not dwell on the format issues.

My expectation is that users will read/write native VSG formats to provide a full and quick means of serializing there scenes, something especially important when doing tasks like database paging where speed is critical.

**Reply by shinyclaw** [reply]
Thanks Robert - I need another fbx converter, because assimp's fbx animation conversion is broken (see https://github.com/assimp/assimp/issues/5449). Again, thanks for the tips!

**Reply by robertosfield** [reply]
The way I'd tackle it would be to implement a vsg::ReaderWriter much like what the vsgchange::assimp does.  You then just an it the vsg::Options readerwriter list, if you put it ahead of the vsgXchange::assimp it'll be called first.
